[venus-gateway]venus-gateway 集成测试(integrate_test目录中)出现的偶发data race错误.

### 链服务模块 / Chain Service Components
- [X] venus-gateway
### 版本 / Version

```text
master分支: 936509d
```


### 描述 / Describe the Bug

venus-gateway集成测试中(integration_test)目录在执行单元测试时添加(-race)检测的时候, 出现数据竞态错误.
`DATA RACE`错误, 此错误通常是由于在并发情况下,对数据进行了没有并发安全控制的读写造成的.
需要更深入的检查, 排除此类错误.

### 日志 / Logging Information
```text
go test -coverpkg=./... -race -coverprofile=coverage.txt -covermode=atomic ./...                                                                                                                                          
WARNING: DATA RACE
Write at 0x00c0008a61f0 by goroutine 26:
  runtime.closechan()
      /usr/local/opt/go/src/runtime/chan.go:356 +0x0
  github.com/ipfs-force-community/venus-gateway/proofevent.(*ProofEventStream).ListenProofEvent.func1.1()
      /Users/zl/workspace/go/src/filcoin-projects/venus/venus-gateway/proofevent/proof_event.go:87 +0x39
  runtime.deferreturn()
      /usr/local/opt/go/src/runtime/panic.go:436 +0x32

Previous read at 0x00c0008a61f0 by goroutine 120:
  runtime.chansend()
      /usr/local/opt/go/src/runtime/chan.go:159 +0x0
  github.com/ipfs-force-community/venus-gateway/types.(*BaseEventStream).sendOnce()
      /Users/zl/workspace/go/src/filcoin-projects/venus/venus-gateway/types/base_event.go:108 +0x3fc
  github.com/ipfs-force-community/venus-gateway/types.(*BaseEventStream).SendRequest()
      /Users/zl/workspace/go/src/filcoin-projects/venus/venus-gateway/types/base_event.go:55 +0x177
  github.com/ipfs-force-community/venus-gateway/proofevent.(*ProofEventStream).ComputeProof()
      /Users/zl/workspace/go/src/filcoin-projects/venus/venus-gateway/proofevent/proof_event.go:125 +0x2eb
  github.com/ipfs-force-community/venus-gateway/api.(*GatewayAPIImpl).ComputeProof()
      /Users/zl/workspace/go/src/filcoin-projects/venus/venus-gateway/api/rpc_impl.go:66 +0x124
  runtime.call128()
      /usr/local/opt/go/src/runtime/asm_amd64.s:704 +0x6a
  reflect.Value.Call()
      /usr/local/opt/go/src/reflect/value.go:339 +0xd7
  github.com/filecoin-project/venus/venus-shared/api/permission.PermissionProxy.func1()
      /Users/zl/workspace/go/pkg/mod/github.com/filecoin-project/venus@v1.6.1-0.20220803064712-14d26a4e39f0/venus-shared/api/permission/permission.go:65 +0x6db
  reflect.callReflect()
      /usr/local/opt/go/src/reflect/value.go:742 +0xaa4
  reflect.callReflect()
      <autogenerated>:1 +0x64
  reflect.makeFuncStub()
      /usr/local/opt/go/src/reflect/asm_amd64.s:47 +0x79
  runtime.call128()
      /usr/local/opt/go/src/runtime/asm_amd64.s:704 +0x6a
  reflect.Value.Call()
      /usr/local/opt/go/src/reflect/value.go:339 +0xd7
  github.com/filecoin-project/go-jsonrpc.doCall()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/handler.go:246 +0x159
  github.com/filecoin-project/go-jsonrpc.(*RPCServer).handle()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/handler.go:352 +0x1b36
  github.com/filecoin-project/go-jsonrpc.(*wsConn).handleCall.func4()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/websocket.go:429 +0x165

Goroutine 26 (running) created at:
  github.com/ipfs-force-community/venus-gateway/proofevent.(*ProofEventStream).ListenProofEvent()
      /Users/zl/workspace/go/src/filcoin-projects/venus/venus-gateway/proofevent/proof_event.go:70 +0x6c9
  github.com/ipfs-force-community/venus-gateway/api.(*GatewayAPIImpl).ListenProofEvent()
      <autogenerated>:1 +0x76
  runtime.call32()
      /usr/local/opt/go/src/runtime/asm_amd64.s:702 +0x48
  reflect.Value.Call()
      /usr/local/opt/go/src/reflect/value.go:339 +0xd7
  github.com/filecoin-project/venus/venus-shared/api/permission.PermissionProxy.func1()
      /Users/zl/workspace/go/pkg/mod/github.com/filecoin-project/venus@v1.6.1-0.20220803064712-14d26a4e39f0/venus-shared/api/permission/permission.go:65 +0x6db
  reflect.callReflect()
      /usr/local/opt/go/src/reflect/value.go:742 +0xaa4
  reflect.callReflect()
      <autogenerated>:1 +0x64
  reflect.makeFuncStub()
      /usr/local/opt/go/src/reflect/asm_amd64.s:47 +0x79
  runtime.call32()
      /usr/local/opt/go/src/runtime/asm_amd64.s:702 +0x48
  reflect.Value.Call()
      /usr/local/opt/go/src/reflect/value.go:339 +0xd7
  github.com/filecoin-project/go-jsonrpc.doCall()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/handler.go:246 +0x159
  github.com/filecoin-project/go-jsonrpc.(*RPCServer).handle()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/handler.go:352 +0x1b36
  github.com/filecoin-project/go-jsonrpc.(*wsConn).handleCall.func4()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/websocket.go:429 +0x165

Goroutine 120 (running) created at:
  github.com/filecoin-project/go-jsonrpc.(*wsConn).handleCall()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/websocket.go:429 +0x7d6
  github.com/filecoin-project/go-jsonrpc.(*wsConn).handleFrame()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/websocket.go:448 +0x1f4
  github.com/filecoin-project/go-jsonrpc.(*wsConn).handleWsConn()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/websocket.go:616 +0xd1b
  github.com/filecoin-project/go-jsonrpc.(*RPCServer).handleWS()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/server.go:80 +0x5f0
  github.com/filecoin-project/go-jsonrpc.(*RPCServer).ServeHTTP()
      /Users/zl/workspace/go/pkg/mod/github.com/ipfs-force-community/go-jsonrpc@v0.1.5/server.go:94 +0x133
  github.com/gorilla/mux.(*Router).ServeHTTP()
      /Users/zl/workspace/go/pkg/mod/github.com/gorilla/mux@v1.8.0/mux.go:210 +0x361
  github.com/filecoin-project/venus-auth/cmd/jwtclient.(*AuthMux).ServeHTTP()
      /Users/zl/workspace/go/pkg/mod/github.com/filecoin-project/venus-auth@v1.6.0/cmd/jwtclient/auth_mux.go:123 +0xd8d
  net/http.serverHandler.ServeHTTP()
      /usr/local/opt/go/src/net/http/server.go:2916 +0x896
  net/http.(*conn).serve()
      /usr/local/opt/go/src/net/http/server.go:1966 +0xbaa
  net/http.(*Server).Serve.func3()
      /usr/local/opt/go/src/net/http/server.go:3071 +0x58
==================
2022-08-18T16:42:38.845+0800	ERROR	rpc	go-jsonrpc@v0.1.5/websocket.go:130	handle me:write tcp 127.0.0.1:57402->127.0.0.1:57403: write: broken pipe
2022-08-18T16:42:38.845+0800	ERROR	rpc	go-jsonrpc@v0.1.5/websocket.go:123	handle me:write tcp 127.0.0.1:57402->127.0.0.1:57403: write: broken pipe
2022-08-18T16:42:38.846+0800	ERROR	rpc	go-jsonrpc@v0.1.5/websocket.go:123	handle me:write tcp 127.0.0.1:57402->127.0.0.1:57403: write: broken pipe
2022-08-18T16:42:38.847+0800	ERROR	rpc	go-jsonrpc@v0.1.5/websocket.go:123	handle me:write tcp 127.0.0.1:57402->127.0.0.1:57403: write: broken pipe
2022-08-18T16:42:38.847+0800	ERROR	rpc	go-jsonrpc@v0.1.5/websocket.go:123	handle me:write tcp 127.0.0.1:57402->127.0.0.1:57403: write: broken pipe
2022-08-18T16:42:38.847+0800	ERROR	rpc	go-jsonrpc@v0.1.5/websocket.go:123	handle me:write tcp 127.0.0.1:57402->127.0.0.1:57403: write: broken pipe
--- FAIL: TestProofAPI (0.03s)
    --- FAIL: TestProofAPI/wait_too_long_and_all_request_timeout (0.01s)
        testing.go:1312: race detected during execution of test
    testing.go:1312: race detected during execution of test
FAIL
coverage: 25.1% of statements in ./...
FAIL	github.com/ipfs-force-community/venus-gateway/integrate_test	0.519s
```


### 重现步骤 / Repo Steps

go test -coverpkg=./... -race -coverprofile=coverage.txt -covermode=atomic ./...

--- a/integrate_test/proofapi_test.go
+++ b/integrate_test/proofapi_test.go
@@ -97,7 +97,7 @@ func TestProofAPI(t *testing.T) {
 		require.NoError(t, err)
 		defer cCloser()
 
-		proofEvent := proofevent.NewProofEvent(walletEventClient, mAddr, &timeoutHandler{}, logging.Logger("test").With())
+		proofEvent := proofevent.NewProofEvent(walletEventClient, mAddr, testhelper.NewTimeoutProofHandler(time.Hour), logging.Logger("test").With())
 		go proofEvent.ListenProofRequest(ctx)
 		proofEvent.WaitReady(ctx)
 
@@ -181,14 +181,6 @@ func TestProofAPI(t *testing.T) {
 		require.Equal(t, 1, len(minerConnections2.Connections))
 		require.Equal(t, 1, minerConnections2.ConnectionCount)
 	})
-}
-
-type timeoutHandler struct {
-}
-
-func (*timeoutHandler) ComputeProof(context.Context, []builtin.ExtendedSectorInfo, abi.PoStRandomness, abi.ChainEpoch, network.Version) ([]builtin.PoStProof, error) {
-	time.Sleep(time.Hour)
-	return nil, nil
 }
 
 func serverProofAPI(ctx context.Context, url, token string) (gateway.IProofEvent, jsonrpc.ClientCloser, error) {

--- a/marketevent/market_event.go
+++ b/marketevent/market_event.go
@@ -61,7 +61,7 @@ func (m *MarketEventStream) ListenMarketEvent(ctx context.Context, policy *types
 	}
 
 	out := make(chan *types2.RequestEvent, m.cfg.RequestQueueSize)
-	channel := types.NewChannelInfo(ip, out)
+	channel := types.NewChannelInfo(ctx, ip, out)
 	mAddr := policy.Miner
 	m.connLk.Lock()
 	var channelStore *channelStore

--- a/proofevent/proof_event.go
+++ b/proofevent/proof_event.go
@@ -57,7 +57,8 @@ func (e *ProofEventStream) ListenProofEvent(ctx context.Context, policy *types2.
 	}
 
 	out := make(chan *types2.RequestEvent, e.cfg.RequestQueueSize)
-	channel := types.NewChannelInfo(ip, out)
+	reqEventChan := make(chan *types2.RequestEvent, e.cfg.RequestQueueSize)
+	channel := types.NewChannelInfo(ctx, ip, reqEventChan)
 	mAddr := policy.MinerAddress
 	e.connLk.Lock()
 	var channelStore *channelStore
@@ -65,6 +66,20 @@ func (e *ProofEventStream) ListenProofEvent(ctx context.Context, policy *types2.
 	if channelStore, ok = e.minerConnections[mAddr]; !ok {
 		channelStore = newChannelStore()
 		e.minerConnections[policy.MinerAddress] = channelStore
+	}
+
+	removeChannel := func() {
+		e.connLk.Lock()
+		channelStore := e.minerConnections[mAddr]
+		e.connLk.Unlock()
+		_ = channelStore.removeChanel(channel)
+		if channelStore.empty() {
+			e.connLk.Lock()
+			delete(e.minerConnections, mAddr)
+			e.connLk.Unlock()
+		}
+
+		log.Infof("remove connections %s of miner %s", channel.ChannelId, mAddr)
 	}
 
 	e.connLk.Unlock()
@@ -75,8 +90,9 @@ func (e *ProofEventStream) ListenProofEvent(ctx context.Context, policy *types2.
 			ChannelId: channel.ChannelId,
 		})
 		if err != nil {
+			removeChannel()
 			close(out)
-			log.Errorf("marshal failed %v", err)
+			log.Errorf("marshal failed %s %v", channel.ChannelId, err)
 			return
 		}
 
@@ -85,26 +101,26 @@ func (e *ProofEventStream) ListenProofEvent(ctx context.Context, policy *types2.
 		stats.Record(ctx, metrics.MinerRegister.M(1))
 		stats.Record(ctx, metrics.MinerSource.M(1))
 
-		out <- &types2.RequestEvent{
+		reqEventChan <- &types2.RequestEvent{
 			ID:         sharedTypes.NewUUID(),
 			Method:     "InitConnect",
 			Payload:    connectBytes,
 			CreateTime: time.Now(),
 			Result:     nil,
 		} // no response
-		defer close(out)
-		<-ctx.Done()
-		e.connLk.Lock()
-		channelStore := e.minerConnections[mAddr]
-		e.connLk.Unlock()
-		_ = channelStore.removeChanel(channel)
-		if channelStore.empty() {
-			e.connLk.Lock()
-			delete(e.minerConnections, mAddr)
-			e.connLk.Unlock()
+
+		// avoid data race
+		for {
+			select {
+			case <-ctx.Done():
+				removeChannel()
+				close(out)
+				stats.Record(ctx, metrics.MinerUnregister.M(1))
+				return
+			case c := <-reqEventChan:
+				out <- c
+			}
 		}
-		log.Infof("remove connections %s of miner %s", channel.ChannelId, mAddr)
-		stats.Record(ctx, metrics.MinerUnregister.M(1))
 	}()
 	return out, nil
 }

--- a/proofevent/proof_event_test.go
+++ b/proofevent/proof_event_test.go
@@ -3,6 +3,7 @@ package proofevent
 import (
 	"encoding/json"
 	"fmt"
+	"sync"
 	"testing"
 	"time"
 
@@ -57,6 +58,39 @@ func TestListenProofEvent(t *testing.T) {
 				return
 			}
 		}
+	})
+
+	t.Run("data race", func(t *testing.T) {
+		proof := setupProofEvent(t, []address.Address{addr1})
+		ctx, cancel := context.WithCancel(context.Background())
+		proofClient := NewProofEvent(proof, addr1, testhelper.NewTimeoutProofHandler(time.Second), log.With())
+		go proofClient.ListenProofRequest(jwtclient.CtxWithTokenLocation(ctx, "127.1.1.1"))
+		proofClient.WaitReady(ctx)
+
+		// fill request
+		wg := sync.WaitGroup{}
+		reqCount := gtypes.DefaultConfig().RequestQueueSize * 2
+		wg.Add(reqCount)
+		for i := 0; i < reqCount; i++ {
+			go func() {
+				wg.Done()
+				_, _ = proof.ComputeProof(context.Background(), addr1, []builtin.ExtendedSectorInfo{}, []byte{}, 100, 16)
+			}()
+		}
+		wg.Wait()
+
+		//cancel and got a close request channel
+		cancel()
+
+		wg2 := sync.WaitGroup{}
+		wg2.Add(reqCount)
+		go func() {
+			for i := 0; i < reqCount; i++ {
+				defer wg2.Done()
+				_, _ = proof.ComputeProof(context.Background(), addr1, []builtin.ExtendedSectorInfo{}, []byte{}, 100, 16)
+			}
+		}()
+		wg2.Wait()
 	})
 
 	t.Run("invalidate address", func(t *testing.T) {

--- a/testhelper/testhelper.go
+++ b/testhelper/testhelper.go
@@ -4,6 +4,7 @@ import (
 	"context"
 	"fmt"
 	"testing"
+	"time"
 
 	"github.com/filecoin-project/specs-storage/storage"
 	sharedTypes "github.com/filecoin-project/venus/venus-shared/types"
@@ -115,4 +116,19 @@ func (p *MarketHandler) SectorsUnsealPiece(ctx context.Context, pieceCid cid.Cid
 		return fmt.Errorf("mock error")
 	}
 	return nil
+}
+
+var _ types.ProofHandler = (*timeoutProofHandler)(nil)
+
+type timeoutProofHandler struct {
+	waitTime time.Duration
+}
+
+func NewTimeoutProofHandler(waitTime time.Duration) types.ProofHandler {
+	return &timeoutProofHandler{waitTime: waitTime}
+}
+
+func (h *timeoutProofHandler) ComputeProof(context.Context, []builtin.ExtendedSectorInfo, abi.PoStRandomness, abi.ChainEpoch, network.Version) ([]builtin.PoStProof, error) {
+	time.Sleep(h.waitTime)
+	return nil, nil
 }

--- a/types/base_event.go
+++ b/types/base_event.go
@@ -7,6 +7,7 @@ import (
 	"fmt"
 	"strings"
 	"sync"
+	"sync/atomic"
 	"time"
 
 	sharedTypes "github.com/filecoin-project/venus/venus-shared/types"
@@ -17,7 +18,7 @@ import (
 
 var log = logging.Logger("gateway_stream")
 
-var ErrCloseChannel = fmt.Errorf("recover send once")
+var ErrCloseChannel = fmt.Errorf("channel closed")
 var ErrRequestTimeout = fmt.Errorf("timer clean this request due to exceed wait time")
 
 type BaseEventStream struct {
@@ -65,11 +66,19 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 	log.Warnf("the first channel is fail, try to other channel")
 	otherChannels := channels[1:]
 	respCh := make(chan *types.ResponseEvent)
+	errRespCount := new(int64)
 	for _, channel := range otherChannels {
 		go func(channel *ChannelInfo) {
 			respEvent, err := e.sendOnce(ctx, channel, method, payload)
 			if err != nil {
+				atomic.AddInt64(errRespCount, 1)
 				log.Errorf("send request %s to %s failed %v", method, channel.Ip, err)
+				// all requests failed
+				if atomic.LoadInt64(errRespCount) == int64(len(otherChannels)) {
+					respCh <- &types.ResponseEvent{
+						Error: fmt.Sprintf("all request failed: %s %s %v", method, channel.Ip, err),
+					}
+				}
 				return
 			}
 			respCh <- respEvent
@@ -85,12 +94,6 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 }
 
 func (e *BaseEventStream) sendOnce(ctx context.Context, channel *ChannelInfo, method string, payload []byte) (response *types.ResponseEvent, err error) {
-	defer func() {
-		if r := recover(); r != nil {
-			err = ErrCloseChannel
-		}
-	}()
-
 	id := sharedTypes.NewUUID()
 	resultCh := make(chan *types.ResponseEvent, 1)
 	request := &types.RequestEvent{
@@ -105,15 +108,20 @@ func (e *BaseEventStream) sendOnce(ctx context.Context, channel *ChannelInfo, me
 	e.reqLk.Unlock()
 
 	select {
-	case channel.OutBound <- request: //NOTICE this may be panic, but will catch by recover and try other, should never have  other panic
-		log.Debug("send request %s to %s", method, channel.Ip)
+	case <-channel.Ctx.Done():
+		return nil, ErrCloseChannel
 	case <-ctx.Done():
 		return nil, fmt.Errorf("send request cancel by context %w", ctx.Err())
+	default:
+		channel.OutBound <- request //NOTICE this may be panic, but will catch by recover and try other, should never have  other panic
+		log.Debug("send request %s to %s", method, channel.Ip)
 	}
 
 	//wait for result
 	//timeout here
 	select {
+	case <-channel.Ctx.Done():
+		return nil, ErrCloseChannel
 	case <-ctx.Done():
 		return nil, fmt.Errorf("cancel by context %w", ctx.Err())
 	case respEvent := <-resultCh:

--- a/types/base_event.go
+++ b/types/base_event.go
@@ -63,11 +63,12 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 
 	//code below unable to work as expect , because there no way to detect network issue in gateway,
 	log.Warnf("the first channel is fail, try to other channel")
+
+	var lk sync.Mutex
+	var respOnce sync.Once
 	otherChannels := channels[1:]
 	respCh := make(chan *types.ResponseEvent)
-	lk := sync.Mutex{}
 	errRespCount := 0
-	respOnce := sync.Once{}
 	for _, channel := range otherChannels {
 		go func(channel *ChannelInfo) {
 			respEvent, err := e.sendOnce(ctx, channel, method, payload)

--- a/types/base_event.go
+++ b/types/base_event.go
@@ -67,6 +67,7 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 	otherChannels := channels[1:]
 	respCh := make(chan *types.ResponseEvent)
 	errRespCount := new(int64)
+	sendRespOnce := sync.Once{}
 	for _, channel := range otherChannels {
 		go func(channel *ChannelInfo) {
 			respEvent, err := e.sendOnce(ctx, channel, method, payload)
@@ -81,7 +82,10 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 				}
 				return
 			}
-			respCh <- respEvent
+			// only send once, avoid goroutine leak
+			sendRespOnce.Do(func() {
+				respCh <- respEvent
+			})
 		}(channel)
 	}
 

--- a/types/base_event.go
+++ b/types/base_event.go
@@ -7,7 +7,6 @@ import (
 	"fmt"
 	"strings"
 	"sync"
-	"sync/atomic"
 	"time"
 
 	sharedTypes "github.com/filecoin-project/venus/venus-shared/types"
@@ -66,20 +65,23 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 	log.Warnf("the first channel is fail, try to other channel")
 	otherChannels := channels[1:]
 	respCh := make(chan *types.ResponseEvent)
-	errRespCount := new(int64)
+	lk := sync.Mutex{}
+	errRespCount := 0
 	respOnce := sync.Once{}
 	for _, channel := range otherChannels {
 		go func(channel *ChannelInfo) {
 			respEvent, err := e.sendOnce(ctx, channel, method, payload)
 			if err != nil {
-				atomic.AddInt64(errRespCount, 1)
-				log.Errorf("send request %s to %s failed %v", method, channel.Ip, err)
+				lk.Lock()
+				errRespCount++
+				log.Errorf("send request %s failed %v", method, err)
 				// all requests failed
-				if atomic.LoadInt64(errRespCount) == int64(len(otherChannels)) {
+				if errRespCount == len(otherChannels) {
 					respCh <- &types.ResponseEvent{
-						Error: fmt.Sprintf("all request failed: %s %s %v", method, channel.Ip, err),
+						Error: fmt.Sprintf("all request failed: %s %v", method, err),
 					}
 				}
+				lk.Unlock()
 				return
 			}
 			// only send once, avoid goroutine leak

--- a/types/base_event.go
+++ b/types/base_event.go
@@ -67,7 +67,7 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 	otherChannels := channels[1:]
 	respCh := make(chan *types.ResponseEvent)
 	errRespCount := new(int64)
-	sendRespOnce := sync.Once{}
+	respOnce := sync.Once{}
 	for _, channel := range otherChannels {
 		go func(channel *ChannelInfo) {
 			respEvent, err := e.sendOnce(ctx, channel, method, payload)
@@ -83,7 +83,7 @@ func (e *BaseEventStream) SendRequest(ctx context.Context, channels []*ChannelIn
 				return
 			}
 			// only send once, avoid goroutine leak
-			sendRespOnce.Do(func() {
+			respOnce.Do(func() {
 				respCh <- respEvent
 			})
 		}(channel)

--- a/types/base_event_test.go
+++ b/types/base_event_test.go
@@ -255,22 +255,27 @@ type mockClient struct {
 
 	closeCh   chan struct{}
 	waitClose chan struct{}
+
+	cancel context.CancelFunc
 }
 
 func setupClient(t *testing.T, event *BaseEventStream, ip string) *mockClient {
 	requestCh := make(chan *types.RequestEvent)
+	ctx, cancel := context.WithCancel(context.Background())
 
 	return &mockClient{
 		t:         t,
 		requestCh: requestCh,
 		event:     event,
-		channel:   NewChannelInfo(ip, requestCh),
+		channel:   NewChannelInfo(ctx, ip, requestCh),
 		closeCh:   make(chan struct{}),
 		waitClose: make(chan struct{}),
+		cancel:    cancel,
 	}
 }
 
 func (m *mockClient) close() {
+	m.cancel()
 	close(m.requestCh)
 	m.closeCh <- struct{}{}
 	<-m.waitClose

--- a/types/base_event_test.go
+++ b/types/base_event_test.go
@@ -239,6 +239,38 @@ func TestSendRequest(t *testing.T) {
 		}
 		eventSteam.reqLk.Unlock()
 	})
+
+	t.Run("all request failed", func(t *testing.T) {
+		ctx, cancel := context.WithCancel(context.Background())
+		defer cancel()
+		eventSteam := NewBaseEventStream(ctx, DefaultConfig())
+
+		parms, err := json.Marshal(mockParams{A: "mock arg"})
+		require.NoError(t, err)
+		result := &mockResult{}
+
+		var clients []*mockClient
+		client := setupClient(t, eventSteam, "127.1.1.1")
+		go client.start(ctx)
+		clients = append(clients, client)
+
+		client2 := setupClient(t, eventSteam, "127.1.1.2")
+		go client2.start(ctx)
+		clients = append(clients, client2)
+
+		var getConns = func() []*ChannelInfo {
+			var channels []*ChannelInfo
+			for _, client := range clients {
+				channels = append(channels, client.channel)
+			}
+			return channels
+		}
+		client.close()
+		client2.close()
+		err = eventSteam.SendRequest(ctx, getConns(), "mock_method", parms, result)
+		require.Error(t, err)
+		require.Contains(t, err.Error(), "all request failed:")
+	})
 }
 
 func TestIstimeOutError(t *testing.T) {

--- a/types/types.go
+++ b/types/types.go
@@ -1,6 +1,7 @@
 package types
 
 import (
+	"context"
 	"time"
 
 	sharedTypes "github.com/filecoin-project/venus/venus-shared/types"
@@ -8,14 +9,16 @@ import (
 )
 
 type ChannelInfo struct {
+	Ctx        context.Context
 	ChannelId  sharedTypes.UUID
 	Ip         string
 	OutBound   chan *types.RequestEvent
 	CreateTime time.Time
 }
 
-func NewChannelInfo(ip string, sendEvents chan *types.RequestEvent) *ChannelInfo {
+func NewChannelInfo(ctx context.Context, ip string, sendEvents chan *types.RequestEvent) *ChannelInfo {
 	return &ChannelInfo{
+		Ctx:        ctx,
 		ChannelId:  sharedTypes.NewUUID(),
 		OutBound:   sendEvents,
 		Ip:         ip,

--- a/walletevent/wallet_event.go
+++ b/walletevent/wallet_event.go
@@ -67,7 +67,7 @@ func (w *WalletEventStream) ListenWalletEvent(ctx context.Context, policy *types
 	ctx, _ = tag.New(ctx, tag.Upsert(metrics.WalletAccountKey, walletAccount), tag.Upsert(metrics.IPKey, ip))
 
 	go func() {
-		channel := types.NewChannelInfo(ip, out)
+		channel := types.NewChannelInfo(ctx, ip, out)
 		defer close(out)
 		//todo validate the account exit or not
 		addrs, err := w.getValidatedAddress(ctx, channel, policy.SignBytes, walletAccount)

